### PR TITLE
Interrupt mode support via idlemgr

### PR DESCRIPTION
**Pull request by KeithWiles**
add idle manager library and test-cne test

The current design in CNDP is 100% polling of the receive ring in AF_XDP
or PMD which means burning core power. The patch adds two parameters to
the jsonc file to control polling idle_timeout and intr_timeout
in thread section.

The idlemgr library handles managing epoll and calling epoll_wait when the
idle_timeout value has been met, then it calls epoll_wait(). The idlemgr
operates on threads, using file descriptors as the method to wakeup a sleeping
thread. The caller needs to add file descriptors to the idlemgr instance and
then call idlemgr_process() with a flag if the thread thinks it is idle or
not.

The first one gives the number of milliseconds to wait for the RX ring to
be idle, then call epoll() with a timeout of intr_timeout. When no Rx traffic
for a given time the idlemgr will call epoll(), which reduces the lcore load to
effectively zero and only waking up when packets arrive or a timeout occurs.
In testing of performance it appears to be very little impact when interrupt
mode is enabled compared to when it is not enabled. Added some counters to help
determine how the new mode is operating.

The idle_timeout value in the jsonc file for given thread is how this feature
is controlled. If not defined or set to zero interrupt mode is disabled. When
set to a non zero value will enable interrupt mode. The intr_timeout value
is only used if idle_timeout is non-zero and will be used in the poll() call
as the timeout value. Each of these values are in milliseconds.

Signed-off-by: Keith Wiles <keith.wiles@intel.com>

### CODE REVIEW

**Comment by KeithWiles** [comment]
This PR replaces #163 by breaking the single commit into multiple commits for easier review. The #163 will be closed.

**Comment by heimafeitian** [comment]
Hi keith，from the review points of ZTE engineer, the intr_timeout and idle_timeout values needs to be added from jsonc lports and lport_groups,because they use the Lport parameter not the thread section.

could you say a little more info about  why removed the intr_timeout and idle_timeout parameters from jsonc lports and lport_groups？

**Comment by KeithWiles** [comment]
The reason is if one thread is managing multiple lports, when one lport becomes idle the thread will go to sleep and starve the other lports. At the xskdev or lport level we can't tell how many lports are managed by the thread, which is why it needed to be moved to the thread level as it contains the lports be handled by the thread.

We also need to be able to manage non-af_xdp lports as well and adding poll() calls in the xskdev driver will not work for all types of lports or multiple lports per thread. If you need the timeout values and not using the thread section the options section could hold the timeout values as this section is mostly user defined.

**Comment by heimafeitian** [comment]
> If you need the timeout values and not using the thread section the options section could hold the timeout values as this section is mostly user defined.

hi keith,
Thank you for your detailed explanation. yes,it is not easy to implement this feature in the lport level.
thanks
Jack

File: lib/core/idlemgr/idlemgr.c [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        /* remove fd from epoll group */
+        if (epoll_ctl(imgr->epoll_fd, EPOLL_CTL_DEL, fd, NULL) < 0)
+            CNE_ERR_GOTO(err_leave, "epoll control delete failed: %s\n", strerror(errno));
+
+        imgr->nb_fds--;
+
+        imgr_unlock(imgr);
+        return 0;
+    }
+    return -1;
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.c:243]:
```suggestion

```

File: lib/core/idlemgr/idlemgr.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ *   The idle_timeout value pointer to get in milliseconds
+ * @param intr
+ *   The intr_timeout value pointer to get in milliseconds
+ * @return
+ *   0 on success or -1 on error
+ */
+CNDP_API int idlemgr_get_timeouts(idlemgr_t *imgr, uint32_t *idle, uint32_t *intr);
+
+/**
+ * Add an file descriptor to the idlemgr instance
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:120]:
```suggestion
 * Add a file descriptor to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+ *   The file descriptor to add to epoll and the idlemgr_t instance
+ * @param eflags
+ *   The flags to be used when setting struct epoll_event.events. If zero use EPOLLIN.
+ * @return
+ *   0 on success or -1 on error
+ */
+CNDP_API int idlemgr_add(idlemgr_t *imgr, int fd, uint32_t eflags);
+
+/**
+ * Delete an file descriptor from the idlemgr instance
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:134]:
```suggestion
 * Delete a file descriptor from the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ * @return
+ *   0 on success or -1 on error
+ */
+CNDP_API int idlemgr_get_timeouts(idlemgr_t *imgr, uint32_t *idle, uint32_t *intr);
+
+/**
+ * Add an file descriptor to the idlemgr instance
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:123]:
```suggestion
 *   The idlemgr_t pointer to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+ * @return
+ *   0 on success or -1 on error
+ */
+CNDP_API int idlemgr_add(idlemgr_t *imgr, int fd, uint32_t eflags);
+
+/**
+ * Delete an file descriptor from the idlemgr instance
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:137]:
```suggestion
  *   The idlemgr_t pointer to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+CNDP_API int idlemgr_del(idlemgr_t *imgr, int fd);
+
+/**
+ * Process the idlemgr instance and call epoll_wait if needed.
+ *
+ * This API needs to be called regularly to ensure that we can process idleness events.
+ * Call this API frequently normally during each RX loop iteration.
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:152]:
```suggestion
 *   The idlemgr_t pointer to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+ *   If nfds is > 0, then it is the number of file descriptors active. Use API idlemgr_get_events
+ *   to grab the struct epoll_event structure list to determine which file descriptors are active.
+ */
+CNDP_API int idlemgr_process(idlemgr_t *imgr, int active);
+
+/**
+ * Return the idle manager struct epoll_event list, after idlemgr_process() returned active FDs.
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:170]:
```suggestion
 *   The idlemgr_t pointer to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+ * @return
+ *   NULL On not found or idlemgr_t pointer if found
+ */
+CNDP_API idlemgr_t *idlemgr_find_by_name(const char *name);
+
+/**
+ * Grab the statistics of the idlemgr_t instance.
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:190]:
```suggestion
 *   The idlemgr_t pointer to the idlemgr instance
```

File: lib/core/idlemgr/idlemgr.h [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+ * @return
+ *   -1 on error or 0 on success
+ */
+CNDP_API int idlemgr_stats(idlemgr_t *imgr, idlemgr_stats_t *stats);
+
+/**
+ * Dump out information about the idlemgr_t instance
+ *
+ * @param imgr
+ *   The idlemgr_t pointer to destroy
```
Comment (maryamtahhan) [inline lib/core/idlemgr/idlemgr.h:202]:
```suggestion
 *   The idlemgr_t pointer to the idlemgr instance
```

File: test/testcne/idlemgr_test.c [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+static int
+test2(void)
+{
+    idlemgr_stats_t stats;
+    idlemgr_t *imgr = NULL;
+    uint32_t idle = 0xFFFFFF, intr = 0xFFFFFF;
+
+    imgr = idlemgr_create("test1", 2, 0, 0);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:80]:
```suggestion
    imgr = idlemgr_create("test2", 2, 0, 0);
```

File: test/testcne/idlemgr_test.c [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    idlemgr_destroy(imgr);
+    imgr = NULL;
+
+    cpu_set_t lcore_bitmap;
+    CPU_ZERO(&lcore_bitmap);
+    CPU_SET(10, &lcore_bitmap);
+    pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &lcore_bitmap);
+    CNE_INFO("LCore ID %d\n", cne_lcore_id());
+
+    imgr = idlemgr_create("test3", 2, 10, 1000);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:126]:
```suggestion
    imgr = idlemgr_create("test2", 2, 10, 1000);
```

File: test/testcne/idlemgr_test.c [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    idlemgr_destroy(imgr);
+    return TST_FAILED;
+}
+
+static int
+test3(void)
+{
+    idlemgr_t *imgr = NULL, *imgr2 = NULL;
+
+    imgr = idlemgr_create("test1", 10, 0, 0);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:175]:
```suggestion
    imgr = idlemgr_create("test3", 10, 0, 0);
```

File: test/testcne/idlemgr_test.c [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+    imgr2 = idlemgr_create("test1", 10, 0, 0);
+    if (imgr2 != NULL)
+        CNE_ERR_GOTO(leave, "idlemgr_create() should fail with duplicate name\n");
+
+    idlemgr_destroy(imgr2);
+    idlemgr_destroy(imgr);
+    imgr = imgr2 = NULL;
+
+    imgr = idlemgr_create("test1", 10, 10, 1000);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:187]:
```suggestion
    imgr = idlemgr_create("test3", 10, 10, 1000);
```

File: test/testcne/idlemgr_test.c [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+    if (imgr == NULL)
+        CNE_ERR_GOTO(leave, "idlemgr_create() succeeded for test1 name\n");
+
+    imgr2 = idlemgr_create("test2", 10, 20, 2000);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:191]:
```suggestion
    imgr2 = idlemgr_create("test3", 10, 20, 2000);
```

File: test/testcne/idlemgr_test.c [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+    if (imgr == NULL)
+        CNE_ERR_GOTO(leave, "idlemgr_create() should succeed with name test2\n");
+    idlemgr_destroy(imgr2);
+
+    imgr2 = idlemgr_create("test2", 10, 30, 3000);
```
Comment (maryamtahhan) [inline test/testcne/idlemgr_test.c:196]:
```suggestion
    imgr2 = idlemgr_create("test3", 10, 30, 3000);
```

**Review by maryamtahhan — APPROVED** [review]
I think one of the commits is missing a sign off but everything else looks good to me

**Comment by KeithWiles** [comment]
Added the signoff to the commit.

**Pull request merged**